### PR TITLE
Add gwctl to gateway-api repo

### PR DESCRIPTION
**Pull request by gauravkghildiyal**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Import `gwctl` packages from https://github.com/gauravkghildiyal/gwctl

**Which issue(s) this PR fixes**:

Fixes #

**Does this PR introduce a user-facing change?**:

```release-note
Add gwctl command-line tool which provides a better way to view and manage policies (https://gateway-api.sigs.k8s.io/geps/gep-713/#kubectl-plugin)
```

### CODE REVIEW

File: gwctl/pkg/policymanager/manager.go
Comment (robscott) [inline gwctl/pkg/policymanager/manager.go:77]: This may benefit from a separate function like `p.AddPolicy(unstructuredPolicy)`

File: gwctl/pkg/policymanager/manager.go
Comment (robscott) [inline gwctl/pkg/policymanager/manager.go:205]:
```suggestion
func PolicyFromUnstructured(u unstructured.Unstructured, policyCRDs map[PolicyCrdID]PolicyCRD) (Policy, error) {
```

File: gwctl/pkg/resources/backends/backends.go
Comment (robscott) [inline gwctl/pkg/resources/backends/backends.go:240]: Non-blocking nit: To simplify the use of this code as a library it may be best to separate code into different packages. For example, one package or set of packages intended to be used as a lib, and a separate package that is more unique to gwctl inputs and outputs (like this describe view here).

I'm not really sure what the ideal code structure is here yet, so this may be worth filing a follow up issue for more discussion once this one merges.

File: gwctl/pkg/resources/policies/policies.go
Comment (robscott) [inline gwctl/pkg/resources/policies/policies.go:40]:
```suggestion
	row := []string{"POLICY NAME", "POLICY KIND", "TARGET NAME", "TARGET KIND"}
```

File: gwctl/pkg/resources/policies/policies.go
Comment (robscott) [inline gwctl/pkg/resources/policies/policies.go:63]:
```suggestion
	row := []string{"NAME", "GROUP", "KIND", "POLICY TYPE", "SCOPE"}
```

File: gwctl/pkg/resources/policies/policies.go
Comment (robscott) [inline gwctl/pkg/resources/policies/policies.go:71]: Would recommend using a string value here of "Inherited" or "Direct" instead

File: gwctl/pkg/types/params.go
Comment (robscott) [inline gwctl/pkg/types/params.go:33]: Nit: I'm not really sure what the best name for this is, but `Params` doesn't seem quite right. It seems like this is a mix of 3 k8s clients, an internal data structure, and an io writer. Are there other ways we can split this up? How broadly is this type used? Are there common combinations that only need a subset of what's in here?

**Review by robscott — COMMENTED** [review]
This is very cool! Thanks for all the work on this @gauravkghildiyal! There's a lot here, but took a high level pass and had some nits + questions.

File: gwctl/pkg/common/consts.go
Comment (gauravkghildiyal) [reply gwctl/pkg/common/consts.go:20]: Nice suggestion. Created https://github.com/kubernetes-sigs/gateway-api/pull/2434 to address this. Will update references once that merges.

File: gwctl/pkg/common/helpers.go
Comment (gauravkghildiyal) [reply gwctl/pkg/common/helpers.go:27]: No that's a good question. I should move the comment from within the function to outside the function. (Will update)

This is just a "good to have" thing which helps provide neat diffs during test. Idea being, split long blobs of yaml text by each line. When the diff is generated, it will clearly point out diffs per line (as oppposed to saying that: "somethings different in this large string, now go find what the difference is")

File: gwctl/pkg/policymanager/manager.go
Comment (gauravkghildiyal) [reply gwctl/pkg/policymanager/manager.go]: Oh no sorry about that. Fixed. Wow how did you catch that amazing.

File: gwctl/pkg/policymanager/manager.go
Comment (gauravkghildiyal) [reply gwctl/pkg/policymanager/manager.go]: Nice idea. Thanks changed.

File: gwctl/pkg/resources/policies/policies.go
Comment (gauravkghildiyal) [reply gwctl/pkg/resources/policies/policies.go]: Great idea. Changed.

File: gwctl/pkg/resources/backends/backends.go
Comment (gauravkghildiyal) [reply gwctl/pkg/resources/backends/backends.go:240]: I definitely resonate with your thoughts here. There's several improvements that could be made to make this more usable as a library. I created https://github.com/kubernetes-sigs/gateway-api/issues/2435 to track the work for this.

File: gwctl/pkg/cmd/describe.go
Comment (gauravkghildiyal) [reply gwctl/pkg/cmd/describe.go:131]: Fair enough. Let me change to just "Services" for now.

File: gwctl/pkg/types/params.go
Comment (gauravkghildiyal) [reply gwctl/pkg/types/params.go:33]: This is definitely a valid commment. I decided to not remain blocked on this initial friction and start with something simple like Params to get moving.
- At the moment, this type is kinda used like a global "context" which is consumed by majority of the functions.
- Initially, I just had three K8s clients here (and hence I initially called it Clients). But as things evolved, and now we have two more things.
- Similar to your other suggestion about `describeViews`, this too definitely requires some code refactor.

I would completely understand the point of view to hold this PR until we fix this (Just say so without any further explanation and I'll be glad to address it here). But I was planning on taking this up separately as part of some code refactoring which is focused. I'll need some time to think the solution here.

File: gwctl/pkg/common/consts.go
Comment (gauravkghildiyal) [reply gwctl/pkg/common/consts.go]: Done and rebased!

File: gwctl/pkg/resources/backends/backends.go
Comment (gauravkghildiyal) [reply gwctl/pkg/resources/backends/backends.go]: Refactored things around. This should be slightly better than the previous iteration. Function names need some refinement but the packages `effectivepolicy` and `policymanager` provide constructs that can be used as a library and are independent from the CLI part.

File: gwctl/pkg/types/params.go
Comment (gauravkghildiyal) [reply gwctl/pkg/types/params.go]: Refactored. Now we have a `CmdParams` but that's limited in scope to only being required by the CLI interface (and not need when library packages are used)

**Comment by robscott** [comment]
> You may consider naming the binary `kubectl-<something>` so its a "kubectl plugin"

Maybe `kubectl-gw` then?

**Comment by robscott** [comment]
Thanks for all the work on this @gauravkghildiyal! I took some time to test this out today and have some high level non-blocking feedback.

1. I'd love to see this be a drop-in replacement for kubectl when reading (get/describe) Gateway API resources, so ideally this would be a superset of kubectl capabilities. With that said, if we end up having to recreate too many things, that's also not ideal. My main point of confusion here is that we're using `get` as a command, but it's not currently a drop-in replacement for `kubectl get`. If we wanted to get there, I think we'd need the following:
    * Support for all resources for both get and describe commands
    * Show at least the same set of columns, ideally more
    * Support for yaml and json output
    * _Maybe_ support for more advanced features like label selection, etc
    * The longer this list gets, the less I'm sure about the overall goal 😄 
2. It would be great to more gracefully handle errors. For example, if one or more Gateway API CRDs are not found, we could just tell the user they need to install the CRD(s) that are missing.
3. Similar to 1, but potentially more achievable in the short term, it would be great to include more information in our describe output. Right now the primary value seems to be policy connections, but it would be useful if it could include that + much of the other useful information included by `kubectl describe`.

I don't think any of these are blockers, but before we go too much further with this, it would probably be helpful to sketch out a user guide for the full set of functionality we want this to be able to accomplish, and then we can split up the work into smaller pieces so it can grow over time from this foundation.

I'm going to go ahead and approve what's here, because it works and looks like a good foundation, but would be very interested in a vision for how we want to build on top of this going forward.

/approve

**Comment by gauravkghildiyal** [comment]
Thanks a lot for the review and great constructive feedback @robscott. Everything that you highlighted is definitely something that I can start addressing ASAP. 

I agree that kubectl doesn't natively provide an extension mechanism in a way in which we can reuse it for certain things which ends up with us having to duplicate things. I will create a user guide on what I think we can/should support so others can start contributing as well.

I had been working on some improvements on the "usage as a library" aspects which I'll send separately.

(Rebasing to resolve merge conflict)

**Review by shaneutt — APPROVED** [review]
Awesome, thanks for adding this!

I think adding it here in the repo is a good idea, as it will keep things in sync.

**Comment by youngnick** [comment]
yeah, I really like this to get us started, but I think we should spend a little time post-GA and talk about what we want this to go - tbh the more I think about it, the more I think we may be better off with two similar but distinct tools - `gwctl` which is a pretty straight `kubectl` replacement for `get` and `describe`, and a separate `gw` kubectl plugin that does the same things but with a slightly different interface. Typing `kubectl gw get all` or something seems slightly more readable than `kubectl gwctl get all`.

/lgtm

**Pull request merged**
